beacon-chain/sync/initial-sync/fsm.go

```
@@ -0,0 +1,232 @@
+func (sm *stateMachine) addHandler(state stateID, event eventID, fn eventHandlerFn) *stateMachineEvent {
+	e, ok := sm.events[event]
+	if !ok {
+		e = &stateMachineEvent{
+			name: event,
+		}
+		sm.events[event] = e
+	}
+	action := &stateMachineAction{
+		state:     state,
+		handlerFn: fn,
+	}
+	e.actions = append(e.actions, action)
+	return e
+}
+
+// trigger invokes the event on a given epoch's state machine.
+func (sm *stateMachine) trigger(name eventID, epoch uint64, data interface{}) error {
+	event, ok := sm.events[name]
+	if !ok {
+		return fmt.Errorf("event not found: %v", name)
+	}
+
+	ind := sm.findEpochState(epoch)
+	if ind >= len(sm.epochs) {
+		return fmt.Errorf("state for %v epoch not found", epoch)
+	}
+
+	for _, action := range event.actions {
+		if action.state != sm.epochs[ind].state {
+			continue
+		}
+		state, err := action.handlerFn(sm.epochs[ind], data)
+		if err != nil {
+			return err
+		}
+		sm.epochs[ind].setState(state)
+	}
+
+	return nil
+}
+
+// addEpochState allocates memory for tracking epoch state.
+func (sm *stateMachine) addEpochState(epoch uint64) {
+	state := &epochState{
+		epoch:   epoch,
+		state:   stateNew,
+		blocks:  make([]*eth.SignedBeaconBlock, 0, allowedBlocksPerSecond),
+		updated: time.Now(),
+	}
+	sm.epochs = append(sm.epochs, state)
+}
+
+// removeEpochState frees memory of processed epoch.
+func (sm *stateMachine) removeEpochState(epoch uint64) error {
+	ind := sm.findEpochState(epoch)
+	if ind >= len(sm.epochs) {
+		return fmt.Errorf("state for %v epoch not found", epoch)
+	}
+	sm.epochs[ind].blocks = nil
+	sm.epochs[ind] = sm.epochs[len(sm.epochs)-1]
+	sm.epochs = sm.epochs[:len(sm.epochs)-1]
+	return nil
+}
+
+// findEpochState returns index at which state.epoch = epoch, or len(epochs) if not found.
+func (sm *stateMachine) findEpochState(epoch uint64) int {
+	for i, state := range sm.epochs {
+		if epoch == state.epoch {
+			return i
+		}
+	}
+	return len(sm.epochs)
```

wouldn't it be better to return an error if its not found ?

---

My reasoning was that when element is not found is not an error, it is normal situation, so it felt kind of wrong to return error when there's no error. Still, since we needed to understand whether item is found or not, I return the length, then one can always do:
```golang
if returnedIndex == len(list) { // we didn't find it
}
```

Do you still advise I change signature to `findEpochState(epoch) (int, error) `?

---

I understand where you are coming from , although if anyone looks at the code initially it will not be immediately obvious that the index was not found. I dont feel too strongly on this, it is fine either way for me.

---

I think you're right, it is not obvious. Will change signature to `findEpochState(epoch) (int, bool)`, so that a conventional `ind, ok := findEpochState(epoch)`, can be used.